solidity/contracts/bridge/VendingMachine.sol

```diff
@@ -0,0 +1,250 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "../token/TBTCToken.sol";
+import "../token/IReceiveApproval.sol";
+import "../GovernanceUtils.sol";
+
+/// @title TBTC v2 Vending Machine
+/// @notice The Vending Machine is the owner of TBTC v2 token and can mint
+///         TBTC v2 tokens in 1:1 ratio from TBTC v1 tokens. TBTC v2 can be
+///         unminted back to TBTC v1 with or without a fee - fee parameter is
+///         controlled by the governance. This implementation acts as a bridge
+///         between TBTC v1 and TBTC v2 token, allowing to mint TBTC v2 before
+///         the system is ready and fully operational without sacrificing any
+///         security guarantees and decentralization of the project.
+///         Vending Machine as a contract itself is not upgradeable, though
+///         TBTC v2 token ownership can be updated in a two-step,
+///         governance-controlled process. It is expected that this process
+///         will be executed before the v2 system launch.
+contract VendingMachine is Ownable, IReceiveApproval {
+    using SafeERC20 for IERC20;
+    using SafeERC20 for TBTCToken;
+
+    /// @notice The time delay that needs to pass between initializing and
+    ///         finalizing update of any governable parameter in this contract.
+    uint256 public constant GOVERNANCE_DELAY = 48 hours;
+
+    /// @notice This divisor for precision purposes. Used to represent fractions
+    ///         in parameter values.
+    uint256 public constant FLOATING_POINT_DIVISOR = 1e18;
+
+    IERC20 public immutable tbtcV1;
+    TBTCToken public immutable tbtcV2;
+
+    /// @notice The fee for unminting TBTC v2 back into TBTC v1. The fee is
+    ///         a portion of the amount being unminted multiplied by
+    ///         `FLOATING_POINT_DIVISOR`. For example, a fee of 1000000000000000
+    ///         means that 0.001 of the value being unminted needs to be paid to
```

Comment:
I do not think saying it is a percentage is a good idea as explained in [this](https://github.com/keep-network/tbtc-v2/pull/9#discussion_r666781701) comment. If we say that it is a percentage, then for 0.01 = 1% fee, I'd expect `unmintFee == 1e18`. I think this would be confusing, it is better to name it a *fraction* IMO.

Reply:
Cool, cool, I just wanted to be clear that this is not a fixed value but is related to the amount that is being unminted.